NFT transfer/minting commands now validate the specified addresses

paninaro opened the pull request:
Some users have been sending NFTs to invalid addresses, such as DID or NFT identifiers. This change adds CLI-side address validation for the NFT transfer and mint commands.

An AddressType enum is introduced with this change which aims to perform bech32m validation of the supplied address, as well as enforcing that the address prefix is of the expected type.

`is_valid_address()` and `ensure_valid_address()` can be called with an `allowed_types` set of `AddressType` enum values. `AddressType.current_network_address_type()` can be called to return the `AddressType` for the config's selected network.

`is_valid_address()` and `ensure_valid_address()` perform the same validation but differ in their return types and exception handling.

Tests included.

paninaro committed:
WIP to introduce a wallet AddressType enum to help validate
user-entered addresses for XCH/DID/NFTs etc.

paninaro committed:
Support grabbing the current network's address prefix while also
allowing the current value to be updated (in case the config changes)

paninaro committed:
Added tests for functionality in/used-by AddressType

paninaro commented:
@Quexington this is essentially the same PR that you've seen before, but I decided to push things further and added the `AddressType` enum. @AmineKhaldi my plan is to replace `DID_HRP` and `NFT_HRP` and replace them with `AddressType.DID/AddressType.NFT` in a follow-up PR.

AmineKhaldi commented:
> @AmineKhaldi my plan is to replace `DID_HRP` and `NFT_HRP` and replace them with `AddressType.DID/AddressType.NFT` in a follow-up PR.

Sounds good, thanks!

Quexington reviewed (commented):
Is there any chance this breaks some functionality when the config is not yet initialized?  It looks like the singleton is set with its initial value as we're loading everything, and that calls a function that reads the config.  Does this break being able to run `chia init`, or am I mistaken?

paninaro commented:
> Is there any chance this breaks some functionality when the config is not yet initialized? It looks like the singleton is set with its initial value as we're loading everything, and that calls a function that reads the config. Does this break being able to run `chia init`, or am I mistaken?

It could be a problem, though not immediately. My hope was to make it as simple as possible to check if an XCH or TXCH address was valid, and have that decision depend on the current config address prefix value (while not having to load the config each time).

I think caching the value will probably end up causing more trouble than its worth, so I'll look for another solution.

altendky reviewed (commented):
It seems that the basic issue underlying all this complexity is the hassle of passing the configuration around through everything that needs to know if we're operating on testnet.  Is it that onerous to make the config available where it is needed?

hoffmang9 commented on chia/wallet/util/address_type.py:
```diff
@@ -0,0 +1,71 @@
+
+
+class AddressType(Enum):
+    XCH = "xch"
+    TXCH = "txch"
+    NFT = "nft"
+    DID = "did:chia:"
```
The issue may more be that it should be tnft and tdid...

altendky replied:
Perhaps.  I would still suggest that those are perhaps not address types but rather an orthogonal dimension.  We are unlikely to hint function `f()` to handle `tnft` and `g()` to handle `nft`.

paninaro replied:
Yes, we should consider making testnet variations for DID and NFT identifiers.

I'll think a bit more about the testnet/mainnet issue. It is a goal to prevent cross-chain address usage (mainnet can't send to TXCH, testnet can't send to XCH), but I take the point about when would we ever use AddressType.TXCH (basically never).

paninaro committed:
Removed code that would automatically resolve the selected network's address prefix.
is_valid_address() and ensure_valid_address() now require the allowed address types to be passed in.

paninaro commented:
> It seems that the basic issue underlying all this complexity is the hassle of passing the configuration around through everything that needs to know if we're operating on testnet. Is it that onerous to make the config available where it is needed?

Yes, that's a fair summary. The latest commit makes the APIs require explicitly setting the allowed address types. I've also added `AddressType.current_network_address_type()` which optionally takes a config or root_path.

Re: "Is it that onerous to make the config available where it is needed?"... I would prefer a model wherein the config is something that can be easily reached from any part of the system without the need to pass it down the call stack.

paninaro commented on chia/wallet/util/address_type.py:
```diff
@@ -0,0 +1,71 @@
+
+    @classmethod
+    def update(cls: Type[_T_CurrentNetworkAddressPrefix], value: str) -> None:
+        cls.current = value
```
Removed

paninaro commented on chia/wallet/util/address_type.py:
```diff
@@ -0,0 +1,71 @@
+
+@default_allowed_types_if_none  # sets allowed_types to xch|txch if not specified
+def ensure_valid_address(address: str, *, allowed_types: Set["AddressType"]) -> str:
+    hrp, data = bech32_decode(address)
```
I could probably add a length dimension to check against. Do you think that's important now? I'm inclined to leave that for a future change if it becomes useful.

altendky replied:
If the not-hrp address is constant length (I think they are?) then it should be an easy add?  Do we have an existing constant that expresses what the length is?  Maybe I'm unaware of something that makes this difficult.

paninaro replied:
Added a length check of the decoded data.

paninaro committed:
Check the length of the decoded address data

altendky commented on tests/core/util/test_config.py:
```diff
@@ -313,3 +303,34 @@ async def test_non_atomic_writes(self, root_path_populated_with_config, default_
                     )
             await asyncio.gather(*all_tasks)
+
+    @pytest.mark.parametrize("prefix", [None])
```
I'm iffy about this passing of parameters to the fixtures like this.  It also seems like a single fixture doing too much.  But, for at least some of this I'll need to dig into it some other time.

altendky reviewed (commented):
It seems like a number of the tests are copy/paste repetitive and could be parametrized instead.

paninaro committed:
Removed TXCH address type. "txch" address validation now requires a config to be passed into one of the validation functions.

altendky commented on chia/wallet/util/address_type.py:
```diff
@@ -0,0 +1,55 @@
+    NFT = "nft"
+    DID = "did:chia:"
+
+    def hrp(self, config: Optional[Dict[str, Any]] = None) -> str:
```
While it may be worthwhile at the moment since we don't have `tnft` etc, making the config optional leaves open errors where even with an xch type you just forget to pass the config and so never get txch.  You can write a bit more code or have to be really super careful any time you touch a bit less code.

I won't require a change here to approve, but be really careful...

paninaro replied:
Agreed that it makes it too easy to make a mistake. The config is now required here and in the validation functions.

altendky commented on tests/conftest.py:
```diff
@@ -608,3 +611,37 @@ async def setup_sim():
+            config["network_overrides"]["config"][config["selected_network"]]["address_prefix"] = prefix
+            save_config(root_path_populated_with_config, "config.yaml", config)
+        updated_config = config
+    return root_path_populated_with_config, updated_config
```
This can just yield the config.  If the test also needs the root path they can just depend on the `tmp_chia_root` directly.  This avoids having to unpack in the test as well as avoiding tuple return values.

paninaro replied:
Renamed and updated so that the fixture yields the config.

altendky reviewed (commented):
Thanks and my apologies for the long-delayed review.  One tidbit on the last fixture.

paninaro committed:
config is now required when using the address validation functions or calling AddressType.hrp().
root_path_and_config_with_address_prefix --> config_with_address_prefix fixture change.

altendky commented on tests/conftest.py:
```diff
@@ -608,3 +611,33 @@ async def setup_sim():
+    return root_path
+
+
+@pytest.fixture(scope="function")
+def config_with_address_prefix(root_path_populated_with_config: Path, prefix: str) -> Iterator[Dict[str, Any]]:
+    updated_config: Dict[str, Any] = {}
+    with lock_and_load_config(root_path_populated_with_config, "config.yaml") as config:
+        if prefix is not None:
+            config["network_overrides"]["config"][config["selected_network"]]["address_prefix"] = prefix
+        yield config
```
Now I have to apologize.  I looked back at my comment and I did use the word yield though I meant it abstractly, not as a Python `yield`...  that was a terrible choice of word.  The issue here is that the config file will be 'locked' in terms of the load functions for the entire test which will probably confuse someone at some point and be a problem.

```suggestion
@pytest.fixture(scope="function")
def config_with_address_prefix(root_path_populated_with_config: Path, prefix: str) -> Dict[str, Any]:
    updated_config: Dict[str, Any] = {}
    with lock_and_load_config(root_path_populated_with_config, "config.yaml") as config:
        if prefix is not None:
            config["network_overrides"]["config"][config["selected_network"]]["address_prefix"] = prefix
    return config
```

altendky commented on tests/conftest.py:
```diff
@@ -7,7 +7,9 @@
-from typing import AsyncIterator, List, Tuple
+from tests.setup_nodes import setup_node_and_wallet, setup_n_nodes, setup_two_nodes
+from pathlib import Path
+from typing import Any, AsyncIterator, Dict, Iterator, List, Tuple
```
```suggestion
from typing import Any, AsyncIterator, Dict, List, Tuple
```

altendky reviewed (commented):
I'm really sorry, this was just me being sloppy in my comment.  `:[`

paninaro committed:
Update tests/conftest.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

paninaro committed:
Update tests/conftest.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>
